### PR TITLE
Add EAGER_BATCHED fetchmode to OneToMany association

### PR DESCRIPTION
**Pull request by woutervanvliet**
In regular EAGER mode, doctrine adds a JOIN for a OneToMany association
forcing the database to load more data in memory than essentially
needed. In a classic Article => Tag concept, each article would be
loaded one extra time for every addition tag.

LAZY mode issues a query for every article to load all associated tags,
also adding more overhead to the database than needed.

This new fetchmode addreses this issue by loading all tags associated to
an article into one additional query.

### CODE REVIEW

File: lib/Doctrine/ORM/Mapping/ClassMetadataInfo.php [hunk 1]
```diff
@@ -161,6 +161,11 @@ class ClassMetadataInfo implements ClassMetadata
      */
     const FETCH_EXTRA_LAZY = 4;
 
+	/**
```
Comment (guilhermeblanco) [inline lib/Doctrine/ORM/Mapping/ClassMetadataInfo.php]: Please do not use tabs

**Comment by EVODelavega** [comment]
Please follow the coding standards: indent with 4 spaces, replace `Array()` with `array()` or `[]`, and put opening brackets for methods/classes on a separate line. (check [doctrine coding standards](https://github.com/deeky666/doctrine-coding-standard/blob/master/Docs/README.md) for details)

File: lib/Doctrine/ORM/UnitOfWork.php [hunk 2]
```diff
@@ -2842,6 +2902,20 @@ public function loadCollection(PersistentCollection $collection)
         $collection->setInitialized(true);
     }
 
+	public function scheduleCollectionForBatchLoading(PersistentCollection $collection) {
```
Comment (EVODelavega) [inline lib/Doctrine/ORM/UnitOfWork.php]: missing doc-block

File: tests/Doctrine/Tests/Models/EagerBatched/Article.php [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+	public function setId($id) {
+		$this->id = $id;
+
+		return $this;
+	}
+
+	/**
+	 * Get title
+	 *
+	 * @return mixed
```
Comment (EVODelavega) [inline tests/Doctrine/Tests/Models/EagerBatched/Article.php]: title is `type="string`, so I'd expect either `@return string` or `@return string|null` if field is nullable. And all methods must conform to PSR-2 (opening bracket on next line)

File: tests/Doctrine/Tests/Models/EagerBatched/Article.php [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+
+	/**
+	 * @Column(type="string")
+	 */
+	protected $title;
+
+	/**
+	 * @OneToMany(targetEntity="Tag",mappedBy="article",cascade={"persist"},fetch="EAGER_BATCHED")
+	 */
+	protected $tags;
```
Comment (EVODelavega) [inline tests/Doctrine/Tests/Models/EagerBatched/Article.php]: Doesn't this require an `ArrayCollection`?

File: tests/Doctrine/Tests/Models/EagerBatched/Tag.php [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+		$me->setArticle($a);
+		$me->setName($tag);
+		return $me;
+	}
+
+	/**
+	 * @ManyToOne(targetEntity="Article",cascade={"persist"},fetch="LAZY",inversedBy="tags")
+	 * @Id()
+	 */
+	protected $article;
```
Comment (EVODelavega) [inline tests/Doctrine/Tests/Models/EagerBatched/Tag.php]: move properties to the top

Comment (interpotential) [inline tests/Doctrine/Tests/Models/EagerBatched/Tag.php]: I don't use it a lot in production code either, but in this case it does make the test code a bit cleaner.

**Comment by interpotential** [comment]
@EVODelavega Thanks! Now I also understand what you meant with your comment on Array, back on Stackoverflow. I've amended as good as possible.

File: tests/Doctrine/Tests/Models/EagerBatched/Article.php [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+     * @Column(type="string")
+     * @var string|null
+     */
+    protected $title;
+
+    /**
+     * @OneToMany(targetEntity="Tag",mappedBy="article",cascade={"persist"},fetch="EAGER_BATCHED")
+     * @var Tag[]|Collection
+     */
+    protected $tags;
```
Comment (EVODelavega) [inline tests/Doctrine/Tests/Models/EagerBatched/Article.php]: `$tags` needs to be initialized in the constructor to a new instance of the `Doctrine\Common\Collections\ArrayCollection` class. See [doctrine documentation](http://doctrine-orm.readthedocs.org/projects/doctrine-orm/en/latest/reference/association-mapping.html#one-to-many-bidirectional) for details

**Comment by EVODelavega** [comment]
I must say that the test-case isn't really a good fit here. I'd expect Articles and Tags to be either a unidirectional one-to-many relation, or a bidirectional many-to-many relation (A collection of tags, zero or more of which can be linked to articles, tags not being linked to articles OR articles being linked to zero or more tags, each tag being linked to one or more articles).
The idea that each tag can have only one article linked to it seems to me to be quite absurd.

This in turn begs the question, if we apply this EAGER_BATCHED mode to many-to-many relation entities, what the possible headaches would be when cascading things like DETACH:

```php
$article1 = $em->find('Article', 1);
$article2 = $em->find('Article', 2);
//if $article2 shares tags with $article1, this might throw an exception:

$em->detach($article1);
//if this tag is shared, and detach cascades, wouldn't the tag instance be unmanaged here?
$article2->getTags()->get(1)->setName('new tag name');
$em->persist($article2);
$em->flush();//if the altered tag is indeed unmanaged, this throws an exception
```

**Comment by interpotential** [comment]
> The idea that each tag can have only one article linked to it seems to me to be quite absurd.

That depends on how one defines Tag. To me, a Tag is a word linked to an Article. The same word may be linked to more articles, in which case it is their own Tag instance. Hence the shared primary key, over $name and $article. And, as you can see, the test case actually creates 100 articles all with the exact same 7 words as tag.

As per the doctrine documentation you linked, one-to-many has to be bidirectional.

**Comment by EVODelavega** [comment]
@interpotential One-to-many doesn't have to be bidirectional (5.6). But aside of that, the current way your entities are set up, you might end up with a series of tags like this:

Name | article
------------ | -------------
foo | 1
foo | 2
foo | 3

These are the exact same words, but the're unique keys (because of the article id). That, to me, seems wasteful. I'd probably use a relational table just storing an article id + tag id like so:

article_id | tag_id
------------ | -------------
1             | 1
1             | 2
2             | 1
2             | 3

To avoid tag duplication, but this is turning into a DB design discussion, instead of a review of a PR

**Comment by interpotential** [comment]
True - though a discussion about DB design is also interesting ;-).

Btw, with all your very useful feedback I'm surprised you overlooked me having misnamed the unit test. Update coming up, once travis-ci has told me that it's ok.

**Comment by Laniax** [comment]
Any updates on this?

**Comment by Koc** [comment]
ping @Ocramius

**Comment by Ocramius** [comment]
@Koc @Laniax I didn't look into this at all, sorry.

**Comment by natebrunette** [comment]
Is there a plan to review this? It would be very useful for us.

**Comment by Ocramius** [comment]
Sorry, I really never got around checking this, but overall my focus is on removing features before we add anything new.

**Comment by antonmedv** [comment]
This is must have feature. It will make everything so easier.

**Comment by beberlei** [comment]
The simplicity of this is beautiful, as it just needs one point in the "hotpath" to hook into to get it working. The additional code to `triggerEagerLoads` looks overly complex to me and use some non-Doctrineisms, but I believe this is something we should look into.

In SQL Alchemy this strategy is called "SELECT IN loading": https://docs.sqlalchemy.org/en/13/orm/loading_relationships.html

**Comment by arthurlauck** [comment]
This would be a amazing functionallity, I used a lot of Eloquent from laravel, and this is a nice way to load relations. Im developing a new applications using doctrine and I stumbled in a problem.

I have a one-to-many relation, but the one is a pagination, so the solution would be use a joined eager load, so it could run in 1 queries like:

`select * from left join product_image product;`

The would be ok, but is hard to cache, if any of the product_image change, I need to invalidate all this collection, in the select in method this would be so much easier.

`select * from product;` and `select * from product_image where id in ({ids})`, so I could separeted in two differentes queries, if any image change, I wouldn't need to invalidate product collection cache, just the product image cache.

**Comment by beberlei** [comment]
Superseded by #8391

**Pull request closed without merging**